### PR TITLE
[flang][OpenMP] Fix LASTPRIVATE for iteration variables

### LINKED ISSUE #64055: [Flang][OpenMP]Incorrect execution result of LASTPRIVATE clause specified in DO construct
```
Version of flang-new : 17.0.0(3ff080b53b0a7f5bccfabd11254e255669892a67)
```

`LASTPRIVATE` clause specified in `DO` construct does not seem to be valid.

omp_PARALLELDO_005i_2.f:
```fortran
      program main
      integer*4 ia, ib(50), ier
      integer*4 :: loop=50
      ia = 0
      ib = 0
!$omp parallel do lastprivate(ia)
      do ia=1,loop
         ib(ia) = ia
      enddo
!$omp end parallel do
      write(*,*) "     ia=",ia,"  It must be ",loop+1
      end
```

```
$ export OMP_NUM_THREADS=2; flang-new -fopenmp omp_PARALLELDO_005i_2.f; ./a.out
      ia= 50   It must be  51
$
```

```
$ export OMP_NUM_THREADS=2; gfortran -fopenmp omp_PARALLELDO_005i_2.f; ./a.out
      ia=          51   It must be           51
$
```

```
$ export OMP_NUM_THREADS=2; ifort -qopenmp omp_PARALLELDO_005i_2.f; ./a.out
      ia=          51   It must be           51
$
```

### PR DESCRIPTION
**Pull request by luporl**
Iteration variables behave slightly different with LASTPRIVATE,
as they must be assigned the value that the copy would have after
sequential execution of the loop. This case is now handled.

This patch also fixes LASTPRIVATE for loops where the step is
different from 1.

Fixes https://github.com/llvm/llvm-project/issues/64055

### CODE REVIEW

File: flang/lib/Lower/OpenMP.cpp [hunk 1]
```diff
@@ -349,34 +354,55 @@ void DataSharingProcessor::insertLastPrivateCompare(mlir::Operation *op) {
...
         //                       omp.wsloop {
         // omp.wsloop {            ...
         //    ...                  store
-        //    store       ===>     %cmp = llvm.icmp "eq" %iv %ub
-        //    omp.yield            fir.if %cmp {
-        // }                         ^%lpv_update_blk:
+        //    store       ===>     %v = arith.addi %iv, %step
+        //    omp.yield            %cmp = %step < 0 ? %v < %ub : %v > %ub
+        // }                       fir.if %cmp {
+        //                           fir.store %v to %loopIV
```
Comment (NimishMishra) [inline flang/lib/Lower/OpenMP.cpp:366]: Nit: An additional `fit.store` here is redundant. We expect all last private operations to be abstracted within `^%lpv_update_blk`

**Review by NimishMishra — COMMENTED** [review]
Thank you for the patch!

If I understand correctly, the previous implementation was generating the `fir.if` depending on a equality check between `v` and `loopIV`. As such, the final update to the iteration variable was being missed.

I'm ok with the approach taken here. Basically, we force the iteration variable to advance by `step`, and then check if it crosses the upperbound `ub`. Depending on whether `step` is positive or negative, a `>` or a `<` predicate is needed for comparison. As such, the ternary select operation seems fine.

I've just one point to raise (@kiranchandramohan can comment better). Through this approach, we have introduced a mandatory set of 1 `arith.addi`, 3 `arith.cmpi`, and 1 `arith.select` in _every_ iteration of the loop. Again, not very sure about this, but can this have a bearing on performance, since loops are extensively present across benchmarks we use for validation? 

Not analysed this in depth, but could the following work?
```
%cmp = arith.cmpi eq, %u, %iv
fir.if %cmp {
    %temp = arith.addi %arg0, %step
     fir.store %temp to %iv
}
```

Basically, we let the current flow (generating `if` on `u = ub`) be, and add the extra operations within the `fir.if`? Would it cause some problems I have missed? I expect if the loop is well formed (say both `ub` and `step` are negative), then also the equality condition should work.

**Review by kiranchandramohan — APPROVED** [review]
Thanks @luporl for the fix. Looks fine to me.

Is the update of the original loop variable unconditionally done now? Should it be only if `lastprivate(loop_variable)` is present? Or is the point that the `loop_variable` is always private and hence the value of the original variable is undefined if it is not `lastprivate` and hence it is OK to update always?

> I've just one point to raise (@kiranchandramohan can comment better). Through this approach, we have introduced a mandatory set of 1 arith.addi, 3 arith.cmpi, and 1 arith.select in every iteration of the loop. Again, not very sure about this, but can this have a bearing on performance, since loops are extensively present across benchmarks we use for validation?

Please refer to https://discourse.llvm.org/t/rfc-privatisation-in-openmp-dialect/3526/4. The `kmpc_for_static_init` calls provide a value called `%.omp.is_last` that can be used to check if this is the last iteration. And updates can be placed by checking this value on the last iteration in the exiting path of the loop. One way to achieve this is by having a new operation in the dialect (in-place of the fir.if and related comparisons) say, `omp.last_private_loop_update`, that takes in the loop_variable, step-size as arguments. This can be extracted out and put in the exit path of the loop. Other alternatives include just modelling the lastprivate variables as part of the omp.wsloop construct and appropriately lowering them with the correct conditions and updates only during lowering to LLVM IR. 

Please wait for @NimishMishra.

**Comment by luporl** [comment]
Thank you for the review @NimishMishra!

> If I understand correctly, the previous implementation was generating the `fir.if` depending on a equality check between `v` and `loopIV`. As such, the final update to the iteration variable was being missed.

Not exactly. Consider this code snippet:

```
integer :: ia
ia = 0
!$omp parallel do lastprivate(ia)
do ia = 1, 10
enddo
!$omp end parallel do
```

In the generated FIR, there will be 3 versions of the `ia` variable:

```
%ia = fir.alloca i32 {bindc_name = "ia", uniq_name = "_QFEia"}
omp.parallel {
    %loopIV = fir.alloca i32 {adapt.valuebyref, pinned}
    omp.wsloop for  (%iv) : i32 = (%lb) to (%ub) inclusive step (%step) {
    }
}
```

`%ia` - the original variable
`%loopIV` - this is the private copy of `%ia`
`%iv` - this is the iteration variable created and managed by `omp.wsloop`

So, the previous implementation was actually checking for equality between `iv` and `ub`. The problem is that `iv` goes from `lb` to `ub`, but LASTPRIVATE expects `ia` to be updated with `iv` + `step`, where `iv`'s value is that of the last loop iteration. `v` was added by this patch and is just the result of `iv` + `step`.
 
> I've just one point to raise (@kiranchandramohan can comment better). Through this approach, we have introduced a mandatory set of 1 `arith.addi`, 3 `arith.cmpi`, and 1 `arith.select` in _every_ iteration of the loop. Again, not very sure about this, but can this have a bearing on performance, since loops are extensively present across benchmarks we use for validation?

It is true that now we need to perform more operations to find out when we are in the last loop iteration. However, these are added just when LASTPRIVATE is used and is optimized to 2 instructions when `lb`, `ub` and `step` are
constant. In the example above, it becomes:
```
%v = add i32 %iv, %step
%cmp = icmp sgt i32 %v, 10
```

But when all of the loop parameters are variables, then I guess it is not always possible to optimize it.

> Not analysed this in depth, but could the following work?
> 
> ```
> %cmp = arith.cmpi eq, %u, %iv
> fir.if %cmp {
>     %temp = arith.addi %arg0, %step
>      fir.store %temp to %iv
> }
> ```
> 
> Basically, we let the current flow (generating `if` on `u = ub`) be, and add the extra operations within the `fir.if`? Would it cause some problems I have missed? I expect if the loop is well formed (say both `ub` and `step` are negative), then also the equality condition should work.

This works only when step is 1 or -1. Consider this loop:

```
ia = 0
!$omp do lastprivate(ia)
do ia = 1, 10, 4
enddo
!$omp end do
```

`iv` will have the following values: 1, 5 and 9. After exiting the loop, `ia` must be updated with 13. With this approach (and without this patch), `iv` will never be equal to `ub`, resulting in `ia` not being updated and remaining with value 0.

File: flang/lib/Lower/OpenMP.cpp (hunk 1 above)
Comment (luporl) [reply flang/lib/Lower/OpenMP.cpp:366]: The `fir.store` here is needed. Considering the example of the previous conversation, `loopIV` is updated in the beginning of the loop body only, to have the same value as `iv`, but we need it to hold `v` instead (`iv` + `step`). Also, `^%lpv_update_blk` copies `loopIV` to `ia` and not `iv` to `loopIV`. In the end this store gets optimized, because `^%lpv_update_blk` generates a load `loopIV` right after it.

**Comment by NimishMishra** [comment]
Thank you for the detailed response. I'm happy with the current changes. Please go ahead with the merge.

**Comment by luporl** [comment]
Thanks for the review @kiranchandramohan!

> Is the update of the original loop variable unconditionally done now? Should it be only if `lastprivate(loop_variable)` is present? Or is the point that the `loop_variable` is always private and hence the value of the original variable is undefined if it is not `lastprivate` and hence it is OK to update always?

The update of the original loop variable is still performed only if `lastprivate(loop_variable)` is present. But the private copy is now being updated, in the last iteration, whenever a `lastprivate` clause is present, even if it doesn't include the loop variable. Do you think it's better to omit the store in this case? I've noticed that this store and the load after it are usually optimized, when llvm bitcode is generated, and only the store of `v` to the original loop variable remains.

> > I've just one point to raise (@kiranchandramohan can comment better). Through this approach, we have introduced a mandatory set of 1 arith.addi, 3 arith.cmpi, and 1 arith.select in every iteration of the loop. Again, not very sure about this, but can this have a bearing on performance, since loops are extensively present across benchmarks we use for validation?
> 
> Please refer to https://discourse.llvm.org/t/rfc-privatisation-in-openmp-dialect/3526/4. The `kmpc_for_static_init` calls provide a value called `%.omp.is_last` that can be used to check if this is the last iteration. And updates can be placed by checking this value on the last iteration in the exiting path of the loop. One way to achieve this is by having a new operation in the dialect (in-place of the fir.if and related comparisons) say, `omp.last_private_loop_update`, that takes in the loop_variable, step-size as arguments. This can be extracted out and put in the exit path of the loop. Other alternatives include just modelling the lastprivate variables as part of the omp.wsloop construct and appropriately lowering them with the correct conditions and updates only during lowering to LLVM IR.

If I understand correctly, this is a suggested optimization for a future patch, correct?
This would be especially useful when support for `lastprivate` with `collapse` is added. Otherwise, we would need to test the loop variables of each nested loop.

**Comment by kiranchandramohan** [comment]
> The update of the original loop variable is still performed only if `lastprivate(loop_variable)` is present. But the private copy is now being updated, in the last iteration, whenever a `lastprivate` clause is present, even if it doesn't include the loop variable. Do you think it's better to omit the store in this case? I've noticed that this store (and the load after it, for loop variables) is usually optimized out, when llvm bitcode is generated, and only the store to the original variable remains.

OK. That is fine.


> If I understand correctly, this is a suggested optimization for a future patch, correct?
> This would be especially useful when support for `lastprivate` with `collapse` is added. Otherwise, we would need to test the loop variables of each nested loop.

Yes, it is something for future work.
